### ISSUE TITLE
Bring Back Kinesis Sources in Platform

### ISSUE THREAD

**Issue by nmeagan11** [MaterializeInc/materialize#14106]
### Feature request

Materialize Platform only supports Kafka and Postgres sources at present. We want to bring back Kinesis sources which were supported in the binary.

Historical Issues to Consider:

- [ ] https://github.com/MaterializeInc/materialize/issues/12304
- [ ] https://github.com/MaterializeInc/materialize/issues/8776
- [ ] Determine what to do about re-sharding (see https://github.com/MaterializeInc/materialize/issues/8995)
- [ ] https://github.com/MaterializeInc/materialize/issues/8764
- [ ] https://github.com/MaterializeInc/materialize/issues/5975
- [ ] https://github.com/MaterializeInc/materialize/issues/5972
- [ ] https://github.com/MaterializeInc/materialize/issues/3251
- [ ] https://github.com/MaterializeInc/materialize/issues/2921
- [ ] https://github.com/MaterializeInc/materialize/issues/2920
- [ ] https://github.com/MaterializeInc/materialize/issues/2895
- [ ] https://github.com/MaterializeInc/materialize/issues/2195

**Comment by benesch** [comment]
@petrosagg, @JLDLaughlin, @sthm, @josharenberg and I discussed this last week. There are two major challenges with Kinesis:

1. Defining the partition ID. Kinesis shards do not work like Kafka partitions, where they have an obvious `PartialOrder` implementation. We'll need to do something more complicated. @petrosagg did some excellent spelunking and came up with the following:

    ```rust
    struct KinesisShard {
      shard_id: u64,
      range_lower: u128,
      range_upper: u128,
    }
   
    struct KinesisTimestamp {
        shard: KinesisShard,
        sequence: [u8; 186], // <- this can probably be improved
    }
    ```

   Then for any two `KinesisTimestamp`s:

    ```
    a < b iff id(a) < id(b) and range(a) overlaps range(b)
    ```

    Details in this Slack conversation: https://materializeinc.slack.com/archives/C01CFKM1QRF/p1694825613419629

2. There's no good Rust SDK for reading from Kinesis. Details here: https://github.com/MaterializeInc/product/issues/175#issuecomment-1722694325

**Comment by benesch** [comment]
Discussion on the other thread reminded me: @petrosagg, is it okay that `ListShards` will eventually stop returning closed shards (i.e., once they transition to expired)? Specifically, as time progresses, it becomes impossible to reconstruct the full history of shard splits and merges if you don't write down the shards as you observe them. I _think_ this is fine (and would save us a lot of work), but wanted to ask explicitly.

**Comment by petrosagg** [comment]
Yes, in fact we only need to call `ListShards` once when we initialize. The moment the source commits its first kinesis frontier in the progress collection we will always know the set of shards that we should resume from and we will be calling `ShardIterator` directly on the outstanding shards

**Comment by petrosagg** [comment]
Of course if we ever resume after a long outage and the shards that we stored in the progress collection have expired under our feet then we will error the source

**Comment by petrosagg** [comment]
This ^ behavior isn't strictly necessary. Another approach would be to call `ListShards` again and jump forward, skipping over an unknown number of messages but that sounds off-brand for Materialize

**Comment by benesch** [comment]
> that sounds off-brand for Materialize

Agreed.
